File: Dynamic-Programming/1218-M-Longest-Arithmetic-Subsequence-of-Given-Difference/longest_arithmetic_subseq.cpp
class Solution {
public:
    int longestSubsequence(vector<int>& arr, int difference) {
        int n = arr.size();
        if (n <= 1) return n;

        unordered_map<int, int> seen;

        int result = 1;
        for (auto num: arr) {
            seen[num] = seen[num - difference] + 1;
            if (result < seen[num]) {
                result = seen[num];
            }
        }
        return result;
    }
};

File: Dynamic-Programming/1218-M-Longest-Arithmetic-Subsequence-of-Given-Difference/longest_arithmetic_subseq.py
import collections
class Solution:
    def longestSubsequence(self, arr: list[int], difference: int) -> int:
        if not arr:
            return 0
            
        seen = collections.Counter()
        for num in arr:
            seen[num] = seen[num - difference] + 1
        return max(seen.values())